### ISSUE TITLE
Introducing 'devel' Branch Workflow

### ISSUE THREAD

**Issue by fantaisie-software** [fantaisie-software/purebasic#84]
As 5.72 is being released soon, I create the 'v5.72' branch. We can still continue to work on master for new features and all bug fixes releated to v5.72 should be merged again in this specific branch. This way we will have historic of all final versions.

**Comment by tajmone** [comment]
The common Git workflow for this type of project is to have a `devel` base branch to merge all approved development features, and possibly also dedicated releases candidates branches (e.g. `devel-572`) or feature specific branches, and then use tagged releases on `master` branch to allow user to checkout specific release snapshots.

I'm strongly against using `master` to merge contributions; it should be used instead to contain just releases commit, which would provide a much cleaner history.

**Comment by SicroAtGit** [comment]
I am also against merging features directly into the `master` branch. There should be a way to unmerge/overwrite a merged feature if a bug/mishap is found later, after the merge, as has happened several times now.

The `master` branch should only receive changes from the `develop` branch, not directly.

The `develop` branch should receive changes from pull requests or directly.

For feature branches or bug fix branches the `master` branch is used as the base.

Releases should be tagged commits from the `master` branch.

**Comment by tajmone** [comment]
What we could do, when the first tagged release is ready, is to rename the `master` branch as `devel` and then create a new orphan `master` branch and squash the tagged release onto it. This would require force pushing and pulling, but since nothing is lost (only renamed) it shouldn't create too many problems for working branches (which should only change their reference).

This way we could achieve a clean history without too much pain.

**Comment by kenmo-pb** [comment]
+1

I'm not a Git expert, but that seems to be how 99% of people recommend to use Git. Keep `master` clean and stable by only including tagged release versions. Use `devel` (or similar) for merging, testing, and fixing contributions.

@tajmone 's idea of turning `master` into `devel` and creating a new, clean `master` branch sounds nice, if we don't mess anything up!

**Comment by fantaisie-software** [comment]
Agreed, let's do this.

**Comment by tajmone** [comment]
> Agreed, let's do this.

To avoid upsetting any contributors currently working on a new feature, we should ensure that everyone is informed about this upcoming change. Also, we should first merge (or close) any pending PRs, for this type of operation could break them.

**Comment by fantaisie-software** [comment]
Makes sense, I will take a look to the remaining PR

**Comment by fantaisie-software** [comment]
I think you can create the devel branch and leave the master as it is. History isn't a big deal IMHO. So from now on (once the devel is created) all PR are merged to devel and then to master if all is OK. When do we decide to merge from dev to master ?

**Comment by tajmone** [comment]
@fantaisie-software:

> I think you can create the `devel` branch and leave the `master` as it is. History isn't a big deal IMHO. 

I agree, it's fine — after all it was clearly stated that the current repository was being transitioned from a privately-run SVN repo to become a public Git repo open to contributions. So the first tagged release is a good point to switch to the new workflow.

> So from now on (once the `devel` is created) all PR are merged to `devel` and then to master if all is OK.

First thing, I'll update the [`CONTRIBUTING.md`][CONTRIBUTING] document to include this info, for it currently mentions that temporarily all commits were going to `master`. I also think you can tweak the repository settings to prevent PRs to `master` branch.

> When do we decide to merge from dev to master?

That's a good question. It really depends on how the workflow is going to be for experimental features — i.e. if there's a baseline `devel` branch were all tested changes are merged into (and from which all contributors PRs should be branched-off from) and maybe also other specialized dev branches for new features (or major changes) that might require prolonged work and should be kept separate from the main workflow.

Usually, Git tags are the way to ensure that users can checkout a specific version of the project (in this case, a specific PB/PB IDE version). But this doesn't necessary mean that between one tagged release and the next one no commits should be made to `master` — after all, some users might just checkout `master` to access the latest features. 

What I think is important is to ensure that only sufficiently tested commits should be merged (or squashed) into `master` — i.e. it should guarantee that all commits contain a usable version of the project; whereas on the `devel` branch there's more room for error and afterthoughts.

Some developers prefer to user `master` only to merge version releases, but I think that using tags is enough to pinpoint releases, and see no harm in further commits adding stable improvements. 

[CONTRIBUTING]: https://github.com/fantaisie-software/purebasic/blob/master/CONTRIBUTING.md#development-cycle-and-strategy

**Comment by tajmone** [comment]
@fantaisie-software, I took the liberty of renaming the Issue since the focus of the discussion shifted toward the adoption of a `devel`-branch based workflow; and also pinned the Issue so it gets more visibility in the repository, which should attract users to join the discussion.

**Comment by tajmone** [comment]
### Live Updates to CONTRIBUTING Document

I've started to update the [`CONTRIBUTING.md`][CONTRIBUTING] document in view of the upcoming `devel` workflow. Although it's still incomplete, I've already created a pull request (PR #91) so it might get reviewed and updated accordingly.

I've added some Git guidelines to help new users adopt the new workflow (e.g. commands to add the `upstream` remote, instructions on how to create a custom devel branch, etc.). These guidelines can still be improved, and surely need proof-reading, but I wanted to start sharing them so they can be discussed.

The idea is that we should provide some insights on how the new workflow works, and also try to make it easy for Git-newbies (after all, the new system does require some extra steps).

The latest edited version of the document can be seen at (links points to the section dealing with Git contributions):

- https://github.com/fantaisie-software/purebasic/blob/devel-instructions/CONTRIBUTING.md#contributing-to-the-repository


For today I'm done, but hopefully you can start viewing the added contents and provide suggestions.

Of course, we are not in a hurry regarding any of this, so it would be nice to ear everyone's view on the upcoming workflow, and ensure that we document it well enough to make it an improvement and not a complication to potential contributors.

[CONTRIBUTING]: https://github.com/fantaisie-software/purebasic/blob/devel-instructions/CONTRIBUTING.md#contributing-to-the-repository "View the CONTRIBUTING document from the PR branch"

**Comment by fantaisie-software** [comment]
Looks great to me

**Comment by tajmone** [comment]
### Merge #91 and Create `devel`

I've finally managed to also update the main README doc, so it mentions that PRs should go to `devel` (sorry for the delay, I was kept away from the PC for two days).

Once PR #91 is merged we should immediately create the new `devel` branch, so the new workflow becomes effective.

As for tagged releases, I think we should wait for the next PB release before creating a tagged release in this repo (especially since currently there's the IDE crash/bug problem of #90).

### Tag-Releases and LTS Updates?

I was wondering, when a PureBasic LTS is officially updated, should this imply a new tagged release for the IDE also? (eg. `572-b`, or something like that).

We haven't discussed how the actual tags should be named, and it's worth thinking about some convention that won't cause problems in the future.

My understanding is that this repository is an umbrella for all the open source components, tools and assets of PureBasic. So it seems to make sense that release tags should mirror the PureBasic edition they match — in case of the IDE, it means that the tag points to the actual source that was used to build the IDE for the official packages.

Even if in the future additional material is added (e.g. the Help docs source files, etc.), it seems reasonable to think that the PB version would still be the main reference.

The only thing which is worth considering is how LTS updates should be coupled to tags, and whether/how to handle bug fixes to the IDE (which might or might not be coupled to an official release?).

Also, usually GitHub tagged releases also offer precompiled binary downloads (besides the sources, which are automatically added). In this context, it probably doesn't make much sense to publish the compiled binaries, because users can download the whole PB package from the official website — that is, unless the repository will contain additional releases (e.g. IDE bug fixes) which are not present in the official releases on the website.

I'm not sure what the pros and cons of all this could be, but probably hosting bug-fix candidate releases here would allow non-contributing users to download the precompiled binaries and test them locally, in order to confirm that the bug was indeed fixed, before creating a new official package (e.g. an LTS update). This is the only pro that comes to my mind, with the downside that it entails manually compiling and adding the IDE binaries at each release, since CI automation is not currently an option (see #3).

**Comment by kenmo-pb** [comment]
I would like to see the `devel` branch finally created...

I am preparing a few Pull Requests (starting with smallest simplest ones) and I'd rather PR into `devel` rather than `master` as we have all discussed before. Keep the `master` branch tidy.

**Comment by tajmone** [comment]
> I would like to see the `devel `branch finally created...

Indeed. It's been ages since the original proposal, but PRs are still going to master.

**Comment by fantaisie-software** [comment]
Go ahead to create the branch, we will merge the last PR not yet merged later on master and all the next one on devel.

**Comment by fantaisie-software** [comment]
Dunno why it was closed but you can go ahead with devel branch. I finished merging all remaining PR

**Comment by tajmone** [comment]
Done. The new `devel` branch in en par with `master`, and there are no pending PRs.

The CONTRIBUTING doc, with instructions to create PRs on devel` branch, was already updated and merged before.

**Comment by fantaisie-software** [comment]
Thank you. Just a few word to say the build system are not yet setup for devel stuff, so I will still push fix on master for now until 5.73 is out. Is there a chat somewhere where we can synchronize ?

**Comment by tajmone** [comment]
> Just a few word to say the build system are not yet setup for devel stuff,

In which sense the build system is not setup for devel? Unless there are branch-specific checks it should work just as fine on any branch.

>  so I will still push fix on master for now until 5.73 is out. 

Makes sense. The new `devel` branch can be used from after the next stable release. If I've understood correctly, we'll be using `devel` as the development baseline branch (from next release onward), and simply merge (without squashing) into `master` whenever a new release is out — meaning that `devel` will be a long-standing branch, and that we might do some force-pushing operations on it if required, in order to preserve a clean commit history. Also, we'll be using sub-branches (i.g. `devel-<some-feature>` for more complex contributions that will require multiple commits and reviews, amending and force-pushing, etc., so that the feature can be squashed into `devel` baseline once it's ready.

> Is there a chat somewhere where we can synchronize ?

I don't regularly use any specific chat system, so can't suggest any ... but I'll join you on any chat service you advise, just provide me with a link.

**Comment by kenmo-pb** [comment]
> Just a few word to say the build system are not yet setup for devel stuff, so I will still push fix on master for now until 5.73 is out.

OK, based on this I will submit a few small things to `master` until 5.73.

**Comment by fantaisie-software** [comment]
This is now over, no more pull request on master. Can you enforce that tajmone ?

**Comment by tajmone** [comment]
>This is now over, no more pull request on master. Can you enforce that tajmone ?

No, you are the only one who can access the branch settings (and other repository related settings).

**Comment by tajmone** [comment]
I've noticed that `devel` branch is lagging behind `master`. Shouldn't it always be en par with it, since it's the dev baseline and needs to be rebasable on `master`? 

My fear is that, being so many commits behind, pull requests might introduce conflicts which are hard to resolve at merge time.

**Comment by fantaisie-software** [comment]
Could you destroy it and re-create it ? nobody used it for it IIRC

**Comment by fantaisie-software** [comment]
Forget it, I just did a rebase master on devel, and switched the default branch to 'devel'. So could yoy tell me if I understood the process right ?
1) Commit/submit new pull request only on 'devel'
2) Ensures it still work
3) Once the devel branch is good and we plan a new release, do a 'rebase devel' on 'master' ? Or cherry-pick all the commits ?
4) Create a branch/tag named about the release or not ?

**Comment by fantaisie-software** [comment]
I just read this post: https://nvie.com/posts/a-successful-git-branching-model/

and it seems very close to what we want to achieve. So we could try to stick to it, we don't reinvent the wheel and use a model which actually work.

**Comment by tajmone** [comment]
> So could yoy tell me if I understood the process right ?

Yes, all the above points are corrects. And, I'd also suggest that in the repo Settings panel you allow only _reabse_ merges, which will ensure that the repo history is always linear when accepting PRs from the WebUI.

Of course, you'd still be free to use other types of merges (and squashing, cherry picking, etc.) via Git, but the above will avoid those anonymous "Merged ..." commit message that are default behaviour on GitHub — but, if I remember correctly, you had already fixed this in the early days.

So, yes, the `devel` branch should always be either en par or ahead of `master`, but always rebasable. If we have some new commits in the `devel` baseline, changes are that it will be ahead of `master`, but right now it's behind (which is not good). Anyhow, I'll rebase it very soon, for I have an upcoming PR that I'd like to create.

> and it seems very close to what we want to achieve. So we could try to stick to it, we don't reinvent the wheel and use a model which actually work.

That article is about Git Flow, which is not a native Git thing but depends on third party add-ons (scripts, etc.) in order to be usable. My Git UI supports it natively, for it's probably the most famous Git workflow in the wild (but there are many others, most of them just agreement based, and simpler to use).

I personally don't like Git flow because it contemplates too many long-standing branches, and matter of fact the Git community is quite divided on this approach, with many people loving it and many other despising it.

It's defintely good for API devlopement, but my understanding is that it introduces "branches hell" — i.e. when you need to create a release you have to integrate all those long-standing branches (bug, devel, features, etc.) and chances are that might have diverged quite a lot, and require a lot of manual chrry picking, rebasing and conflict solving to actually merge into a release.

The minimum I can say is that it requires all participants to master Git heavily, and be very disciplined in the way they handle branches, and especially commits and their messages, so that conflicts can be easily traced and solved. It's usually mostly used amongst closed-nit collaborative teams that have planned out very well their dev strategy ahead of time, and where everyone has a task and knows what others are doing.

Also, since it requires some set up, it might be a detterant for the occasional contributor, who'd have to setup the Gitflow scripts, which might not be supported by his/her Git GUI (in case he/she uses one).

I tend to stick to simpler worflows, like the "Cactus model":

- https://barro.github.io/2016/02/a-succesful-git-branching-model-considered-harmful/

But probably here we need to keep things flexible, for the project really contains lots of different beasts (the IDE, the docs, etc.), most of which might just need a short-standing branch from `devel` and then be merged back, but others might indeed be longer-standing feature that might require longer time (e.g. migrating formats or tools), so we can't really enforce a fixed model.

As long as the `devel` baseline is kept clean and `master` can rebase on it we should be fine to go. Longer standing dedicated branches should follow the same principle, and if conflicts arise they can be solved soon enough if these branches are constantly rebased on their baseline.

**Comment by tajmone** [comment]
OK, I've rebased `devel` on `master` and force-pushed it, so now it can be used as a baseline for new PRs, and I can start pushing the `Documentation/` fixes I've worked on so far...

**Comment by fantaisie-software** [comment]
OK, let's do like that and see what happen
